File: app/src/main/java/com/yawar/memo/modelView/ConversationModelView.java
package com.yawar.memo.modelView;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.yawar.memo.model.ChatMessage;
import com.yawar.memo.model.ChatRoomModel;
import com.yawar.memo.repositry.BlockUserRepo;
import com.yawar.memo.repositry.ChatMessageRepo;
import com.yawar.memo.repositry.ChatRoomRepo;
import com.yawar.memo.utils.BaseApp;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.Objects;

public class ConversationModelView extends ViewModel {
        BaseApp baseApp = BaseApp.getInstance();
        private final ChatMessageRepo repository = baseApp.getChatMessageRepo();
        private  final BlockUserRepo blockUserRepo = baseApp.getBlockUserRepo();
        private String lastSeen = "null";


    private   String _state;
    public MutableLiveData<String> state ;
    private   String _isTyping;
    public MutableLiveData<String> isTyping ;
    public MutableLiveData<Boolean> isFirst ;





    public ConversationModelView() {

        this.state=new MutableLiveData<>("false");
        this.isTyping = new MutableLiveData<>();
        this.isFirst = new MutableLiveData<>(true);

    }


//
    public void setBlockedFor(String blockedFor) {
        blockUserRepo.setBlockedForRepo(blockedFor);
    }
    public MutableLiveData<String> blockedFor(){
        System.out.println("block for changeeeeeee");
        return blockUserRepo.blockedForRepo;
    }

    public MutableLiveData<Boolean> isBlocked(){
        return blockUserRepo.blockedRepo;
    }
    public void setBlocked(Boolean blocked){
        blockUserRepo.setBlockedRepo(blocked);
    }

    public MutableLiveData<Boolean> isUnBlocked(){
        return blockUserRepo.getUnBlockedRepo();
    }
    public void setUnBlocked(Boolean unBlocked){
        blockUserRepo.setUnBlockedRepo(unBlocked);
    }

    public void sendBlockRequest(String my_id, String another_user_id){
        blockUserRepo.sendBlockRequest(my_id,another_user_id);
    }
    public void sendUnBlockRequest(String my_id, String another_user_id){
        blockUserRepo.sendUnbBlockUser(my_id,another_user_id);
    }


    public MutableLiveData<ArrayList<ChatMessage>> getChatMessaheHistory() {
            return repository.chatMessageistMutableLiveData;

        }



    public void addSelectedMessage(ChatMessage message){
        repository.addSelectedMessage(message);
//        _selectedMessage.add(message);
//        selectedMessage.setValue(_selectedMessage);
    }
    public void removeSelectedMessage(ChatMessage message){
        repository.removeSelectedMessage(message);
//        _selectedMessage.remove(message);
//        selectedMessage.setValue(_selectedMessage);
    }
    public void clearSelectedMessage(){
        repository.clearSelectedMessage();
//        repository.clearMessageChecked();

    }
    public MutableLiveData<ArrayList<ChatMessage>> getSelectedMessage() {
        return repository.selectedMessage;

    }





    public void addMessage(ChatMessage message) {
            repository.addMessage(message);
        }

        public void deleteMessageFromList(JSONArray jsonArray) {
            try {


                for (int i = 0; i < jsonArray.length(); i++) {

                    String message_id = jsonArray.getString(i);
                    for (ChatMessage chatMessage : Objects.requireNonNull(getChatMessaheHistory().getValue())) {
                        if (chatMessage.getId().equals(message_id)) {
                            System.out.println(chatMessage.getId() + " " + message_id);
                            repository.deleteMessage(chatMessage);
                            break;
                        }
                    }


                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    public void deleteMessage( ) {
        for (int i = 0; i < getSelectedMessage().getValue().size(); i++) {

            String message_id = getSelectedMessage().getValue().get(i).getId();
            for (ChatMessage chatMessage : Objects.requireNonNull(getChatMessaheHistory().getValue())) {
                if (chatMessage.getId().equals(message_id)) {
                    System.out.println(chatMessage.getId() + " " + message_id);
                    repository.deleteMessage(chatMessage);
                    break;
                }
            }


        }
    }
    public void ubdateMessage( String messge_id,String message) {
        repository.UpdateMessage(messge_id,message);
    }
    public void setMessageChecked( String messge_id,boolean isChecked) {
        repository.setMessageChecked(messge_id,isChecked);
    }
    public void setMessageState( String messge_id,String messageState) {
        repository.setMessageState(messge_id,messageState);
    }



    public void set_state(String _state) {
        System.out.println("stateTooles"+_state);
        this._state = _state;
        state.setValue(_state);
    }
    public void set_isTyping(String _isTyping) {
        this._isTyping = _isTyping;
        isTyping.setValue(_isTyping);
    }

    public void setMessageDownload(String message_id,boolean isDwonload) {
        repository.setMessageDownload(message_id,isDwonload);
    }

    public void deleteMessageForMe(ArrayList<String> message_id,String user_id) {

         repository.deleteMessageForMe(message_id.toString(),user_id,getSelectedMessage().getValue());




    }

    public String getLastSeen() {
        return lastSeen;
    }

    public void setLastSeen(String lastSeen) {
        this.lastSeen = lastSeen;
    }

    public MutableLiveData<Boolean> getLoading(){
        return repository.loading;
    }

    public MutableLiveData<Boolean> getErrorMessage(){
        return repository.showErrorMessage;
    }
    public void setLoading(Boolean check){
        repository.loading.setValue(check);
    }
    public void setErrorMessage(Boolean check){
        repository.showErrorMessage.setValue(check);
    }


}

File: app/src/main/java/com/yawar/memo/modelView/ConversationModelView.kt
package com.yawar.memo.modelView
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.yawar.memo.model.ChatMessage
import com.yawar.memo.utils.BaseApp
import org.json.JSONArray
import org.json.JSONException
import java.util.*

class ConversationModelView : ViewModel() {
    var baseApp = BaseApp.getInstance()
    private val chatMessageRepoo = baseApp.chatMessageRepoo
    private val blockUserRepo = baseApp.blockUserRepo
    var lastSeen = "null"




    private val _state =  MutableLiveData<String>("false")
    val state : LiveData<String>
        get() = _state



    private val _isFirst =  MutableLiveData<Boolean>(true)
    val isFirst : LiveData<Boolean>
        get() = _isFirst

    private val _isTyping =  MutableLiveData<String>()
    val isTyping : LiveData<String>
        get() = _isTyping




    //
    fun setBlockedFor(blockedFor: String?) {
        if (blockedFor != null) {
            blockUserRepo.setBlockedForRepo(blockedFor)
        }
    }

    fun blockedFor(): LiveData<String> {
        return blockUserRepo.blockedForRepo
    }

    val isBlocked: LiveData<Boolean>
        get() = blockUserRepo.blockedRepo

    fun setBlocked(blocked: Boolean) {
        blockUserRepo.setBlockedRepo(blocked)
    }

    val isUnBlocked: LiveData<Boolean>
        get() = blockUserRepo.unBlockedRepo

    fun setUnBlocked(unBlocked: Boolean) {
        blockUserRepo.setUnBlockedRepo(unBlocked)
    }

    fun sendBlockRequest(my_id: String, another_user_id: String) {
        blockUserRepo.sendBlockRequest(my_id, another_user_id)
    }

    fun sendUnBlockRequest(my_id: String, another_user_id: String) {
        blockUserRepo.sendUnbBlockUser(my_id, another_user_id)
    }
    fun getChatMessaheHistory(): LiveData<ArrayList<ChatMessage?>?> {
//            return repository.chatMessageistMutableLiveData;
        return chatMessageRepoo.chatMessaheHistory
    }

    fun addSelectedMessage(message: ChatMessage?) {
        chatMessageRepoo.addSelectedMessage(message)
        //        _selectedMessage.add(message);
//        selectedMessage.setValue(_selectedMessage);
    }

    fun removeSelectedMessage(message: ChatMessage?) {
        chatMessageRepoo.removeSelectedMessage(message!!)
        //        _selectedMessage.remove(message);
//        selectedMessage.setValue(_selectedMessage);
    }

    fun clearSelectedMessage() {
        chatMessageRepoo.clearSelectedMessage()
        //        repository.clearMessageChecked();
    }

    val selectedMessage: LiveData<ArrayList<ChatMessage?>?>
        get() = chatMessageRepoo.selectedMessage

    fun addMessage(message: ChatMessage?) {
        chatMessageRepoo.addMessage(message!!)
    }

    fun deleteMessageFromList(jsonArray: JSONArray) {
        try {
            var chatMessageList = chatMessageRepoo.chatMessaheHistory.value

            for (i in 0 .. jsonArray.length()) {

                val message_id = jsonArray.getString(i)
                if (chatMessageList != null) {
                    for (chatMessage in chatMessageList) {
                        if (chatMessage != null) {
                            if (chatMessage.id == message_id) {
                                chatMessageRepoo.deleteMessage(chatMessage)
                                break
                            }
                        }
                    }
                }
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
    }

//    fun deleteMessage() {
//        for (i in selectedMessage.value.indices) {
//            val message_id = selectedMessage.value!![i]!!.id
//            for (chatMessage in Objects.requireNonNull(chatMessaheHistory.getValue())) {
//                if (chatMessage.id == message_id) {
//                    println(chatMessage.id + " " + message_id)
//                    chatMessageRepoo.deleteMessage(chatMessage)
//                    break
//                }
//            }
//        }
//    }

    fun ubdateMessage(messge_id: String?, message: String?) {
        chatMessageRepoo.UpdateMessage(messge_id!!, message)
    }

    fun setMessageChecked(messge_id: String?, isChecked: Boolean) {
        chatMessageRepoo.setMessageChecked(messge_id!!, isChecked)
    }

    fun setMessageState(messge_id: String?, messageState: String?) {
        chatMessageRepoo.setMessageState(messge_id!!, messageState!!)
    }

    fun set_state(state: String) {
        _state.value = state
    }

    fun set_isTyping(isTyping: String) {
        _isTyping.value = isTyping
    }

    fun setMessageDownload(message_id: String?, isDwonload: Boolean) {
        chatMessageRepoo.setMessageDownload(message_id!!, isDwonload)
    }

    fun deleteMessageForMe(message_id: ArrayList<String?>, user_id: String?) {
        chatMessageRepoo.deleteMessageForMe(message_id.toString(),
            user_id,
            selectedMessage.value)
    }

    fun setLoading(value: Boolean) {
        chatMessageRepoo.setLoading(value)
    }

    fun setErrorMessage(value: Boolean) {
        chatMessageRepoo.setErrorMessage(value)
    }
    fun setState(value: String) {
       _state.value = value
    }
    fun setIsFirst(value: Boolean) {
        _isFirst.value = value
    }
    fun setTyping(value: String) {
        _isTyping.value = value
    }

    fun  getLoading() :LiveData<Boolean>{
        return chatMessageRepoo.loading;
    }
    fun  getErrorMessage() :LiveData<Boolean>{
        return chatMessageRepoo.errorMessage;
    }


}
